File: 백준/Gold/13023. ABCDE/ABCDE.cc
#include<iostream>
#include<list>

using namespace std;

int N, M;
int ck[2000] = { 0, };
list<int>f[2000];
int cnt = 1;
int flag = 0;

void DFS(int num) {
	if (cnt == 5) {
		flag = 1;
		return;
	}
	ck[num] = 1;
	
	for (auto it : f[num]) {
		if (ck[it] == 0) {
			cnt++;
			DFS(it);
			cnt--;
		}
	}

	ck[num] = 0;
}

int main() {

	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	cin >> N >> M;

	int a, b;
	for (int i = 0; i < M; i++) {
		cin >> a >> b;
		f[a].push_back(b);
		f[b].push_back(a);
	}

	for (int i = 0; i < N; i++) {
		DFS(i);
		if (flag)break;
	}
	

	cout << flag << "\n";

	return 0;
}

File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {

	public static StringBuilder sb = new StringBuilder();
	public static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
	public static StringTokenizer st;

	static int N, M;

	static List<List<Integer>> p = new ArrayList<>();
	static boolean visit[] = new boolean[2000];

	static boolean flag = false;

	static void DFS(int idx, int cnt) {
		if (cnt == 5) {
			flag = true;
			return;
		}
		visit[idx] = true;
		for (int i : p.get(idx)) {
			if (!visit[i]) {
				DFS(i, cnt + 1);
			}
		}
		visit[idx] = false;
	}

	public static void main(String[] args) throws IOException {

		st = new StringTokenizer(in.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		for (int i = 0; i < N; i++) {
			p.add(new ArrayList<>());
		}

		for (int i = 0; i < M; i++) {
			int a, b;
			st = new StringTokenizer(in.readLine());

			a = Integer.parseInt(st.nextToken());
			b = Integer.parseInt(st.nextToken());

			p.get(a).add(b);
			p.get(b).add(a);
		}

		for (int i = 0; i < N; i++) {
			DFS(i, 1);
			if (flag) {
				System.out.println(1);
				return;
			}
		}

		System.out.println(0);
	}
}